### PR TITLE
Move push-ci-build.sh functionality to release repo.

### PR DESCRIPTION
**Pull request by david-mcmahon**
cc @ixdy @spxtr 

ref https://github.com/kubernetes/kubernetes/pull/24629
ref https://github.com/kubernetes/kubernetes/issues/16529

### CODE REVIEW

**Comment by ixdy** [comment]
Is this ready for review?

**Comment by david-mcmahon** [comment]
Yes, I'd like to push this soon to test out the new release-based push mechanism.

**Comment by david-mcmahon** [comment]
@ixdy @spxtr Yes, if you could review this now.  Please focus on what might be missing when run from within a Jenkins context.   ref https://github.com/kubernetes/test-infra/pull/279

**Comment by david-mcmahon** [comment]
ping.  If you can just give this a quick check, that should be fine.  I expect to iterate on this, but I need to get something in to feed initial jenkins job.

File: README-CI.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Continuous Integration How-to
+
+Called from hack/jenkins/build.sh
```
Comment (ixdy) [inline README-CI.md]: I assume another PR is going to do that?

File: lib/releaselib.sh [hunk 2]
```diff
@@ -198,7 +198,7 @@ release::set_build_version () {
       return 1
     else
       logecho
-      logecho "Bad build version!"
+      logecho "Bad build version: $build_version"
```
Comment (ixdy) [inline lib/releaselib.sh]: isn't `$build_version` only set in the passing if branch?

**Comment by ixdy** [comment]
cc @zmerlynn too

File: README-CI.md (hunk 1 above)
Comment (david-mcmahon) [inline README-CI.md]: https://github.com/kubernetes/kubernetes/pull/28922

File: lib/releaselib.sh (hunk 2 above)
Comment (david-mcmahon) [inline lib/releaselib.sh]: Yes, fixed in next commit.

File: lib/releaselib.sh [hunk 3]
```diff
@@ -638,19 +661,28 @@ release::gcs::verify_release_gt() {
...
-      greater=false
+      greater=0
     elif [[ -z "$version_prerelease" && -n "$gcs_version_prerelease" ]]; then
       : # fall out
     elif [[ "$version_prerelease" < "$gcs_version_prerelease" ]]; then
-      greater=false
+      greater=0
     elif [[ "$version_prerelease" > "$gcs_version_prerelease" ]]; then
       : # fall out
     # Finally resort to -le here, since we want strictly-greater-than.
```
Comment (ixdy) [inline lib/releaselib.sh]: comment may need to be updated?

File: lib/releaselib.sh [hunk 4]
```diff
@@ -638,19 +661,28 @@ release::gcs::verify_release_gt() {
...
-    elif [[ "$version_prerelease_rev" -le "$gcs_version_prerelease_rev" ]]; then
-      greater=false
+    elif [[ $publish_file =~ ^release && \
```
Comment (ixdy) [inline lib/releaselib.sh]: why are you special-casing on release vs. ci only here?

File: lib/releaselib.sh [hunk 5]
```diff
@@ -584,16 +605,16 @@ release::gcs::publish_official () {
 # @param version - release version
 # @return 1 if new version is not greater than the GCS version
 #
-release::gcs::verify_release_gt() {
+release::gcs::verify_latest_update () {
```
Comment (ixdy) [inline lib/releaselib.sh]: tangential note: this method really wants unit tests.

File: push-ci-build.sh [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+###############################################################################
+# MAIN
+###############################################################################
+RELEASE_BUCKET=${FLAGS_bucket:-"kubernetes-release-dev"}
+RELEASE_BUCKET_MIRROR=$FLAGS_bucket_mirror
+# We're calling directly into kubernetes/kubernetes here so we have to use
+# these globals.  Later these should be passed as command-line flags to
+# local vars
+#: ${FEDERATION_PUSH_REPO_BASE:="gcr.io/k8s-jkns-e2e-gce-federation"}
```
Comment (ixdy) [inline push-ci-build.sh]: why are these commented out?

File: anago [hunk 7]
```diff
@@ -952,17 +952,21 @@ for label in ${!RELEASE_VERSION[@]}; do
   common::stepheader "PUSH ${RELEASE_VERSION[$label]} IMAGES"
   ##############################################################################
   common::runstep release::gcs::copy_release_artifacts \
```
Comment (ixdy) [inline anago]: minor API nit: it's a little unfortunate that `copy_release_artifacts` uses `TYPE/VERSION` as one argument whereas `publish_version` uses `TYPE VERSION` as two separate arguments.

**Comment by david-mcmahon** [comment]
Thanks for the comments.  All addressed in some form. Agree on the unit test for release::gcs::verify_latest_update().  TODO for the moment.  Plan to get some in shortly.

**Comment by david-mcmahon** [comment]
@ixdy PTAL.  Unit test added for release::gcs::verify_latest_update().

**Comment by david-mcmahon** [comment]
If this is otherwise fine and good enough for testing the [test yaml](https://github.com/kubernetes/test-infra/pull/279) I'm going to merge this sometime this week.

**Comment by ixdy** [comment]
LGTM

**Pull request merged**